### DISCUSSION TITLE
VSCode

### DISCUSSION THREAD

**Discussion by jhechtf** [standardagents/arrow-js#23] [General]
I'm willing to look into this myself, but I am wondering if there is a way to make a VSCode extension for arrow so that it will highlight HTML inside the `html` or `t` tagged template literals. Wanted to keep the discussion somewhere and this didn't particularly seem like an issue as it was more of a curiosity.

**Comment by jhechtf** [comment]
Okay so today was a weird day at work, so managed to spend some time. There does _seem_ like there is a way, but I'm not 100% on how to start on it. Seems like we would be using embedded language support, but i need to look more into how it works. Might build a simple POC to verify.

**Reply by justin-schroeder** [reply]
Good question. I would love to have more fine grained control over this too with a first party VSCode extension, buuuut...the lit-html vscode extension does do this well for the "html" tag

**Reply by jhechtf** [reply]
I think, and again, would need to fudge about with this more, that we could do some injections into the base language (html) and then use the embedded language with those injections.

**Comment by jhechtf** [comment]
okay so since lit-html extension has a couple of grammars ive borrowed them and stuck them into bare-bones extension. So far, seems to work. What sort of specialization would we look for in this case?

![image](https://user-images.githubusercontent.com/8152471/205468168-68a7db64-6913-4d7d-9333-a1826385516f.png)

**Comment by jhechtf** [comment]
If anyone wants to take a look at this very WIP extension, it's over at https://github.com/jhechtf/vscode-arrow-js

**Reply by justin-schroeder** [reply]
@jhechtf would love to take a look! I think the repo is private though?

**Reply by jhechtf** [reply]
that's weird. give me a second. I thought I had already put it to public

**Reply by jhechtf** [reply]
okay, try it now @justin-schroeder

**Reply by Chaphasilor** [reply]
I just took a look, this is so much nicer than the regular template strings! 😭  
Thank you for your efforts, please don't abandon them! I'm using arrow for multiple projects already, and it's awesome to work with (aside from some bugs), so I would love to see better "syntax" support.
Never took a look at `lit-html`, but I'm guessing it's mostly the same.

What I noticed is that the "arrow function" construct is highlighted in a strange way sometimes: 
![image](https://user-images.githubusercontent.com/18015147/209847340-93dcb580-69ab-41c4-8cbf-bdee5e538881.png)
![image](https://user-images.githubusercontent.com/18015147/209847378-e28b733b-ab9e-45d4-ac6a-a08d5718c48b.png)

Other times it works better:
![image](https://user-images.githubusercontent.com/18015147/209847460-48fdcec6-b632-4b22-b77f-dfd67d2aacea.png)
![image](https://user-images.githubusercontent.com/18015147/209847520-621f969f-ce24-43f6-8262-73785c949b51.png)

In regular JavaScript however the whole construct seems to have the same color:
**Edit:** maybe that's just a coincidence, the color of the parentheses probably depends on the nesting level.
![image](https://user-images.githubusercontent.com/18015147/209847671-48fee8c6-c393-4492-b79b-c2e011694ca3.png)
![image](https://user-images.githubusercontent.com/18015147/209847707-89d929d4-cc7e-4b81-8aee-c4fb5d3c991a.png)
![image](https://user-images.githubusercontent.com/18015147/209847747-35d580db-48a9-495c-80d7-9a26995afa40.png)

Not sure if you could do anything about this? Of course it's not a deal-breaker, but I think the goal is to have the best and most consisten syntax highlighting possible, right? Otherwise we could just keep using lit-html :)

**Reply by Chaphasilor** [reply]
Oh and one more thing, so far this seems to only be syntax highlighting, but it's not actually telling VS Code that the section within the `html` template string is a different language, right? That would be awesome!  
This actually happens in regular html every time you open a `script` or `style` tag, when IntelliSense suggests different things. Same for `.vue` single-file components. Maybe you want to look into how that's done? 😁

**Reply by jhechtf** [reply]
Hmm, likely something with the grammar since im "borrowing" the grammer built for Lit. never really dealt with language grammars before so it's all been a bit of a learning experience.
